File: triangle.ts
{
    function minimumTotal(triangle: number[][]): number {
        const dp = Array(triangle.length + 1).fill(0);

        for (let i = triangle.length - 1; i >= 0; i--) {
            const row = triangle[i];

            for (let j = 0; j < row.length; j++) {
                dp[j] = row[j] + Math.min(dp[j], dp[j + 1]);
            }
        }
        console.log(dp)
        return dp[0];
    };

    console.log(minimumTotal([
            [-1], 
            [2], 
           [2,3], 
          [1,-1, 3]
    ]))
}

File: triangle.rs
pub fn minimum_total(mut triangle: Vec<Vec<i32>>) -> i32 {
    for y in 1..triangle.len() {
        for x in 0..=y {
            triangle[y][x] += if x == 0 {
                triangle[y - 1][x]
            } else if x == y {
                triangle[y - 1][x - 1]
            } else {
                triangle[y - 1][x].min(triangle[y - 1][x - 1])
            }
        }
    }
    *triangle.last().unwrap().iter().min().unwrap()
}
